wheatley/tower.py

```diff
@@ -216,69 +212,66 @@ def _on_user_leave(self, data):
         # Unassign all instances of that user
         for bell, user in self._assigned_users.items():
             if user == user_id_that_left:
-                self._assigned_users[bell] = None
+                del self._assigned_users[bell]
 
                 bells_unassigned.append(bell)
 
         self.logger.info(
             f"RECEIVED: User #{user_id_that_left}:'{user_name_that_left}' left from bells {bells_unassigned}."
         )
 
-    def _on_user_entered(self, data):
+    def _on_user_entered(self, data: JSON) -> None:
         """ Called when the server receives a uew user so we can update our user list. """
         # Add the new user to the user list, so we can match up their ID with their username
         self._user_name_map[data['user_id']] = data['username']
 
-    def _on_user_list(self, user_list):
+    def _on_user_list(self, user_list: JSON) -> None:
         """ Called when the server broadcasts a user list when Wheatley joins a tower. """
         for user in user_list['user_list']:
             self._user_name_map[user['user_id']] = user['username']
 
-    def _join_tower(self):
+    def _join_tower(self) -> None:
         """ Joins the tower as an anonymous user. """
         self._emit(
             "c_join",
             {"anonymous_user": True, "tower_id": self.tower_id},
             f"Joining tower {self.tower_id}"
         )
 
-    def _request_global_state(self):
+    def _request_global_state(self) -> None:
         """ Send a request to the server to get the current state of the tower. """
         self._emit('c_request_global_state', {"tower_id": self.tower_id}, "Request state")
 
-    def _emit(self, event: str, data, message: str):
-        """ Emit a socket-io signal. """
-        if self._socket_io_client is None:
-            raise Exception("Not Connected")
-
-        self._socket_io_client.emit(event, data)
-
-        if message:
-            self.logger.info(f"EMIT: {message}")
-
-    def _on_bell_rung(self, data):
+    def _on_bell_rung(self, data: JSON) -> None:
         """ Callback called when the client receives a signal that a bell has been rung. """
         self._update_bell_state(data["global_bell_state"])
```

Comment:
This also needs to convert from JSON or push the `bool` -> `Stroke` conversion into `self._update_bell_state`

Reply:
Oh yes good spot.  I wish that Python had strongly typed JSON (de)serialisation which would catch errors like this, but as it stands I think I'll to a thorough check through `tower.py` and scrutinise all the type conversions.

Reply:
Actually, I think we could enforce strong typing by unpacking the values into variables and then putting types on those (with casts if necessary).

Reply:
What do you think of making a (local) function for each out-going SocketIO signal, so that we can enforce correct types for those signals too (and automatically catch errors like this)?

i.e. something like:
```python
def emit_bell_rung(bell_num: int, is_handstroke: bool):  
    self._emit(
        "c_bell_rung",
        {"bell": bell_num, "stroke": is_handstroke, "tower_id": self.tower_id},
        ""
    )

emit_bell_rung(bell.number, stroke.is_hand())
```

Reply:
@centreboard, what do you think about the above message?

I think this is the only thing blocking the merging of this PR...

Reply:
It's a nice idea, but I think we'd end up with lots of duplicate methods. Lots of our emit methods are small.

I think we can ignore `self.tower_id` as it would break everything and maybe extract others into local variables to make us think twice about changing them?

```python
emitted_bell: int = bell.number
emitted_stroke: bool = stroke.is_hand()
self._emit(
    "c_bell_rung",
    {"bell": emitted_bell, "stroke": emitted_stroke, "tower_id": self.tower_id},
    ""
)
```

Reply:
That sounds like a better compromise.  I'll do that now.

Reply:
Turns out that this was the only place that the types weren't declared elsewhere, so it changing this was easier than I expected...